src/accelerate/data_loader.py

```diff
@@ -294,12 +296,26 @@ def __init__(self, dataset, device=None, rng_types=None, generator=None, **kwarg
         self.device = device
         self.rng_types = rng_types
         self.generator = generator
+        self.gradient_state = GradientState()
 
     def __iter__(self):
         if self.rng_types is not None:
             synchronize_rng_states(self.rng_types, self.generator)
+        self.gradient_state._set_end_of_dataloader(False)
+        _current_batch = 0
+        _dataloader_length = len(self) - 1 if hasattr(self, "__len__") else False
         for batch in super().__iter__():
+            if _dataloader_length:
+                self.gradient_state._set_end_of_dataloader(_current_batch == _dataloader_length)
+                _current_batch += 1
+            else:
+                self.gradient_state._set_end_of_dataloader(False)
             yield batch if self.device is None else send_to_device(batch, self.device)
+        if not self.gradient_state.end_of_dataloader:
+            self.gradient_state._set_end_of_dataloader(True)
+            logger.warn(
+                "Warning! DataLoader had no length and finished iterating. Backwards pass and stepping must be manually performed one last time."
+            )
```

Comment:
This is unnecessary complex. Why are you not doing `self.gradient_state._set_end_of_dataloader(True)` at the end of `iter`?

Reply:
This check has to happen as the last batch is being pushed out, otherwise it doesn't occur until outside the iteration loop. 

Example in python of what doesn't happen:

```python
a = [0,1,2,3]
class A:
    def __init__(self,a):
        self.a,self.is_end = a,False
    def __iter__(self):
        for item in a: yield item
        self.is_end = True
aa = A(a)
for i,item in enumerate(aa):
    print(f'At end at {i}?: {aa.is_end}')
print(aa.is_end)
```
Result:
```
At end at 0?: False
At end at 1?: False
At end at 2?: False
At end at 3?: False
True
```
What we want to have happen:
```python
class A:
    def __init__(self,a):
        self.a,self.is_end = a,False
    def __iter__(self):
        for i,item in enumerate(a):
            if i == len(a)-1:
                self.is_end = True 
            yield item
        self.is_end = True
```
```
At end at 0?: False
At end at 1?: False
At end at 2?: False
At end at 3?: True
True
```
Since we cannot check the length, this extra step + warning needs to be done.

Reply:
I think there's a way to do this easier, will toy with it today.